### ISSUE TITLE
How to handle the deprecation of dynamic properties in PHP 8.2 and PHP 9.0

### ISSUE THREAD

**Issue by daigo75** [woothemes/woocommerce#31316]
Is your feature request related to a problem? Please describe.
--
The PHP development team (i.e. one guy) announced that, starting from PHP 8.2, dynamic properties will be deprecated, and then removed altogether (see https://wiki.php.net/rfc/deprecate_dynamic_properties). Dynamic properties are used in a fairly large number of projects, and, admittedly, I've been using them extensively as a simple form of state tracking. 

In many of my projects, I need to process a set of objects (e.g. products). All object instances must be processed once, and only once. If two identical objects come along, then each one must be processed once, and only once. The simple example below illustrates such a scenario:
```
$product1 = wc_get_product(123);
$product2 = wc_get_product(123);

// These calls must process each product
do_action('some_action', $product1);
do_action('some_action', $product2);

// These calls must not do anything
do_action('some_action', $product1);
do_action('some_action', $product2);
```
This tracking is required due to the architecture of WordPress and WooCommerce. It's not possible to predict if an action, or filter is going to be triggered more than once, therefore it's essential to keep track of the state of each object instance. 

The challenge arises from the fact that the various instances of objects are absolutely identical, i.e. they don't have a unique ID, nor properties that could be used to identify uniquely. Due to that, using dynamic properties is a simple solution to this. Example:
```
add_action('some_action', function($product) {
  // Property "some_state_property" doesn't exist in the original class, it's "tacked"
  // to the specific instance
  if(!isset($product->some_state_property)) {
    // Process the product
    $product->some_state_property = 'some tracking value';
  }
  return $product;
}
```
This mechanism allows to track an object state, without interfering with any of its built-in features. However, without dynamic properties, this logic will break.

Describe the solution you'd like
--
**The easiest solution** to this challenge would be to keep allowing dynamic properties for the WooCommerce objects. The documentation indicates that this is possible by using the #[AllowDynamicProperties] tag before the class. This is inherited by descendant classes, therefore the solution would be as simple as modifying the base abstracts. For example:
```
#[AllowDynamicProperties]
abstract class WC_Data {
  // Code here...
}
```
This would allow dynamic properties on all the classes descending from WC_Data, which would basically mean all data-related objects (products, orders, etc). That would allow all the existing code, which leverages dynamic properties, to keep working as it does.

Describe alternatives you've considered
--
Without dynamic properties, we would need a different mechanism to identify and "tag" each object instance individually. Methods like `WC_Data::add_meta_data()` are not suitable because, by design, they always cause the meta to be saved to the database if `WC_Data::save()` is called. This must not happen with state data. Object state data is temporary, and it's used for the processing of an object instance. Due to that:
 1. It must be specific to each instance of an object.
 2. It must not be saved to the database.
 3. It must not be used by other instances of the same object.

Due to this, the only alternatives to dynamic properties would be the following:
1. Having a way to set instance-specific data, not to be saved in the database.
2. Having a way to identify an object instance uniquely, e.g. [with a GUID](https://de.wikipedia.org/wiki/Globally_Unique_Identifier). The GUID would have to be generated when an object is instantiated, and it would make that instance uniquely identifiable. Note: functions like [spl_object_id()](https://www.php.net/manual/en/function.spl-object-id.php) are not suitable. In typical PHP fashion, these functions recycle the "unique" IDs, and can assign them to different objects. Due to that, they may return an ID that belonged to another object, leading to potential nightmare in figuring out with the data matches the wrong object.

Additional context
--
The tracking of an object state is essential in all those cases where each object must be processed only once. In my case, the most common scenario would be dealing with prices, which could be converted depending on the currency, discounted depending on customer's role or country. In these scenarios, the logic that applies the conversion, or a discount, must only do that once per product instance. 

Here's an example of such scenario (simplified, for illustration purposes):
```
add_filter('woocommerce_product_get_price', function($price, $product) {
	// The hypothetical function user_has_discount() checks if the customer is
	// entitled to a 10% discount.
	// Property $product->already_discounted is dynamic, and added it to the 
	// product instance after the discount has been applied. This prevents
	// the repeated application of the discount
  if(user_has_discount() && !isset($product->already_discounted)) {
		$product->set_price($price * 0.9);
		// Tag the product as "already discounted", so the discount won't be applied again
		$product->already_discounted = true;
	}

	return $price;
}, 10, 2);

// Load two instances of the same product. They are identical to each other
$product1 = wc_get_product(123);
$product2 = wc_get_product(123);

// Output the discounted price
echo "YOUR PRICE FOR PRODUCT 1: " . $product1->get_price();
echo "YOUR PRICE FOR PRODUCT 2: " . $product2->get_price();

// Output the discounted price again. Thanks to the 
// safeguard in the filter, the price will still be the same, 
// no matter how ofter WC_Product::get_price() is called
echo "YOUR PRICE FOR PRODUCT 1: " . $product1->get_price();
echo "YOUR PRICE FOR PRODUCT 2: " . $product2->get_price();
```

Conclusion
--
Without the check in the filter, multiple calls to `WC_Product::get_price()` would result in the discount being applied multiple times. It's therefore essential to keep track of the state of each instance, so that each one can be processed only when needed. Dynamic properties are a convenient way to do that. Unlike values set with `add_meta_data()`, these properties reside only in memory, and they are forgotten when the object instance is dropped, which is exactly what should happen.

An alternative solution would be to track the state separately from the original object (e.g. in some sort of "state dictionary"), but then it would be imperative that each object instance can be identified uniquely, without any chance that its ID could be "recycled" for other objects (hence the suggestion of a GUID for each instance).

**Comment by peterfabian** [comment]
Thanks for raising this, @daigo75 and for the great explanation. I think this deserves a bit more in-depth research, as just enabling this across all our classes might have implications we don't know about just yet. 

Additionally, I think there are other ways to achieve a similar effect without necessarily using dynamic properties, from simple ones like having an array/map to keep track of processed objects to hooking filters per object instance and unhooking them as soon as the function finishes (which is basically the same, just with extra steps, the list of filters would keep track of which objects would still get processed). Let's keep this issue open to discuss possible approaches and implications of the dynamic properties setting/flag in the context of WC.

**Comment by daigo75** [comment]
@peterfabian I can already comment on this:
> I think this deserves a bit more in-depth research, as just enabling this across all our classes might have implications we don't know about just yet.

In short, no, it won't. It will make the code work exactly as it does now. Hence my suggestion of that being a quick fix.

The other solution you describe, of using an external dictionary, is also one of the options I evaluated (which, I realised, I didn't describe in in my post, but I definitely evaluated). However, keeping an external list can only work if one can distinguish, **with absolute certainty**, one object instance from the other. That is, not ID of the entity that the object represents, the ID of the instance itself.

From my post:
> Due to this, the only alternatives to dynamic properties would be the following:
> 1. Having a way to set instance-specific data, not to be saved in the database.
> 2. **Having a way to identify an object instance uniquely, e.g. with a GUID**. The GUID would have to be generated when an object is instantiated, and it would make that instance uniquely identifiable. Note: functions like spl_object_id() are not suitable. In typical PHP fashion, these functions recycle the "unique" IDs, and can assign them to different objects. Due to that, they may return an ID that belonged to another object, leading to potential nightmare in figuring out with the data matches the wrong object.

If you had some kind of `get_guid()` method in the base classes, that would also work fine. It's a very simple implementation, which can rely on lazy intialisation. Example (using Traits, for even easier coding):
``` 
trait Instance_GUID {
  /**
   * Stores the unique ID of the object instance.
   *
   * @var string
   */
  protected $_guid = null;

  /**
   * Returns the unique ID of the object instance.
   * 
   * @return string;
   */
  public function get_guid(): string {
    // Function generate_guid() would return a unique ID.
    // @link https://www.uuidgenerator.net/dev-corner/php
    return $this->_guid ?? ($this->_guid = generate_guid());
  }
}

class WC_Data {
  use Instance_GUID;
}
```

This would have several advantages:
1. No side effect on any other element. The `get_guid()` is  a new method, so nothing else is using it and can't possibly affect other elements.
2. Very low performance impact. If nobody needs the GUID (e.g. "base" WooCommerce), the method is not going to be called, thus it would be almost as if it didn't exist. If someone does need the GUID, the generation is called only once.

This seems a simple solution to me, quick and easy to implement, without side effects. It would just be a matter of adding the function to the appropriate base classes.

Additional notes
--
I'm not sure I understand the following:
> keep track of processed objects to hooking filters per object instance and unhooking them as soon as the function finishes

However, the hook/unhook won't work in most of my scenarios. In my case, I need to be able to process an object once **per context**. If the context changes, the object must be processed again. Here's a practical example, just for illustration purposes, which involves multi-currency prices (the bane of many developers 😅):
```
// Change the active currency
set_active_currency('USD');
// This call should process the object and return the USD price
$price = $product->get_price();
// This call should not do processing, just return the price
$price = $product->get_price();

set_active_currency('EUR');
// This call should process the object again and return the EUR price
$price = $product->get_price();
// This call should not do processing, just return the price
$price = $product->get_price();
```

The above is one the most common scenario I have to deal with. **All** the calls to the `get_price()` method must be intercepted (so the "remove hook after processing" is not possible), but only **some** of them will trigger a processing. Due to that, my tracking is not just a simple "processed/not processed", but "processed for currency X" or "processed for currency Y".

**Comment by barryhughes** [comment]
> However, keeping an external list can only work if one can distinguish, with absolute certainty, one object instance from the other.

Is this not possible with a regular array and strict comparison? There's no need to worry about the object ID, which has the limitations you noted. 

Even better, there are dedicated types you might use—such as [SplObjectStorage](https://www.php.net/manual/en/class.splobjectstorage.php) and perhaps better yet again (especially from a memory efficiency perspective) there is the [WeakMap](https://www.php.net/manual/en/class.weakmap.php) (though this is only available in PHP 8.0 and above: but since both implement `ArrayAccess` you could conceivably make a decision on which to use at runtime).

**Comment by daigo75** [comment]
@barryhughes I'm not sure what you mean with a "_a regular array and strict comparison_". The problem is not storing objects, it's that there must be a way to distinguish the state of separate object instances. Currently, there is no attribute that can uniquely identify an instance, hence my suggestion to use a unique ID with lazy initialisation.

The two proposed alternatives are not bad, but they have some important drawbacks:
1. The `SplObjectStorage` has the drawback of increasing the object count, thus increasing the lifespan of an object. That can hinder performance and increase the memory footprint. I can't predict how many objects will be passed to the processing logic, it could be hundreds for a single page call.
2. The `WeakMap` would be a better solution, and its purpose is precisely what I would be implementing myself:
"_Its primary use case is for building caches of data derived from an object that do not need to live longer than the object._"
However, it's only for PHP 8. 

Ideally, I would like to find a solution that I can implement right now (currently, the minimum requirement PHP 7.1, so nothing over that) and not have to revisit later. The existing logic has been working for 8 years straight, so I would be aiming at a similar "implement and forget" solution. Admittedly, I'm not a fan of implementing the same logic in multiple ways (additional testing, multiple versions), unless I have no choice. This time there is still plenty of time to make a choice, hence my early report. 😁

Digression: needless to say, I voted against the breaking change in PHP 8.2/PHP 9, and I would have left the deprecation as "opt-in" only. Also, From a purely management perspective, I would never have allowed a developer to add breaking changes a week before leaving the project. That might be because I come from a background of mission-critical projects, where backward incompatible changes are "bad by default", and last minute changes even more so. 🤷‍♂️

**Comment by barryhughes** [comment]
> I'm not sure what you mean with a "a regular array and strict comparison". 

I mean if you place a reference to an object in an array, and your callback subsequently receives a reference to the _same_ object, you can verify that it is the same via [strict comparison of the objects](https://www.php.net/manual/en/language.oop5.object-comparison.php) (so, at a high level, a similar idea to using the dedicated [SplObjectStorage](https://www.php.net/manual/en/class.splobjectstorage.php) type).

> That can hinder performance and increase the memory footprint.

Yes, that would be a drawback in some scenarios.

> The above is one the most common scenario I have to deal with.

Coming back to the example you shared (set active currency to USD, then set to some other currency, etc) I'm still a little unsure why this would necessitate using dynamic properties. You talked of 'processing': so are you doing something other than filtering on `woocommerce_product_get_price` (are you actually mutating each instance)?

**Comment by daigo75** [comment]
I'm not sure that comparing whole objects could be efficient. It seems like a performance bottleneck, especially since we're talking about complex objects, such as `WC_Product` instances. 

As for the following:
> are you doing something other than filtering on woocommerce_product_get_price (are you actually mutating each instance)?

I'm doing both. I'm using `woocommerce_product_get_price` to alter the prices of a product. This is called by calling `WC_Product::set_price()`. However, this is not relevant to this conversation, as the subject is tracking the changes on a per-instance basis. The logic I'm using is solid, tested and it's going to stay the way it is. The only question I need to answer is how I can preserve the performance optimisations I have already implemented, by using a different way to track an object state.

**Additional notes**
In case the question about the use of `woocommerce_product_get_price` filter was related to the fact that "_one should not change the object properties, but only returned filtered values_", I can save some time: I already heard that. Leaving aside the fact that I don't necessarily agree with it, that idea might sound valid in theory, but not in practice (exactly like the removal of dynamic properties 😄). I already tried that approach, ever since the whole CRUD logic was introduced, and it causes all sorts of conflicts with other plugins. 

I won't spend too much time explaining all the details of the conflicts, both because this off-topic, and because I already spent dozens of hours on it back in 2016, but I can summarise it as follows:
1. You can check a dozen random plugins and you will find that basically all of them alter the product of a price by calling `WC_Product::set_price()`.  I've also been told that "_nobody should do that_", but they do, so we all have to deal with it.
2. Due to the above, returning the modified price from the filter, without altering the instance, won't work, because it will ignore the changes made by other plugins, that do not return a price through the filter.
3. There's no issue in altering the instance of a product and setting its prices. As I mentioned a couple of times already, the only shortcoming of the current design is that the methods used to change the object always assume that the changes should be persisted to the database, when that's not necessarily the case. However, there are ways to prevent unwanted side effects. 

Everything I discussed so far has been working fine for almost a decade. I've been working with WooCommerce since its version 1.6.x, so I believe I have sufficient experience to know what I'm doing. 😁
As soon as I can find a way to implement a reliable, alternative way of tracking the object states, it will keep working fine like it did until now. 

Conclusion
--
The target of this discussion is not to review the specifics of my price conversion logic. I pointed out the introduction of a breaking change, which just so happens to affect my code. From my side, I just need to find an effective solution to the state tracking approach that I've been using so far (which, to be frank, I also found a bit unelegant, but it works and it's fast) that would not cause side effects. 

Right now, I can see that we discussed at least six possible solutions: the three options that I proposed initially (keep dynamic properties, use in-memory attributes, introduce a GUID), which, in my opinion, are easy to implement and without apparent side effects, plus the other three discussed today (object comparison, using `SplObjectStorage`, using `WeakMap`). which could also be viable, but present limitations such as memory consumption, slower performance, incompatibility with previous PHP versions. To be quite frank, these last three options would not be my first choice (hence my other suggestions).

I believe that, by now, we have covered the subject sufficiently, and gathered enough data and options to make a decision in the coming months. It should not be a matter of urgency, as [PHP 8.2 should be released at the end of 2022](https://php.watch/versions). I will leave this to you, so that you can discuss it internally and choose an approach. I will review my logic as well, and add a note if I find a quick solution.

**Comment by barryhughes** [comment]
> The target of this discussion is not to review the specifics of my price conversion logic.

:100: ... but it is important to us to have a good understanding of the use cases requiring the change, and this was a nice vehicle for that (so, thank you for all the detailed responses).

> I'm not sure that comparing whole objects could be efficient. It seems like a performance bottleneck, especially since we're talking about complex objects, such as WC_Product instances.

I would have thought the opposite: specifically in the case of a strict comparison (a check as to whether the two variables refer to the same instance) I'd have thought it would be extremely fast for the runtime to do this, *irrespective* of how complex or simple the objects being compared might be. That said, whether I'm right or wrong about that, there are other drawbacks you already noted (re object lifetime).

> I will leave this to you, so that you can discuss it internally and choose an approach.

We'll do that (and, again, I appreciate all the notes).

**Comment by daigo75** [comment]
That was no bother at all. Mine is a specific use case, but I also worked on at least a dozen of other projects (not necessarily products) where there was a need to "tag" objects being processed. These objects could be duplicated, but they would not offer any `set_meta()` method, nor allow any custom attributes, nor properties. Just like we discussed, using the `SplObjectStorage`, or an array to track specific instances, would have caused memory issues, and `WeakMap` didn't exist, so dynamic properties were an excellent "patch" to that.

Here's a very rough example of this technique, which I've used countless times:
~~~
foreach($a_ton_of_objects as $object) {
  // Do something with $object 
  // Track the progress
  if(<SOME CONDITION>) {
    $object->custom_state = 'state1';
  }
  else {
    $object->custom_state = 'state2';
  }
}

// Use the custom state later on
if($object->custom_state === 'state1') {
   do_next_step($object);
}
~~~
This works right away, has no side effects (as long as the dynamic properties are named properly) and doesn't require using separate dictionaries or collections. When the object "dies", so do also the dynamic properties. They are like a label sticking to them. 

Tacking properties to an object feels a bit "dirty", but it works. In the wild world of JavaScript, it's actually a technique that I see all the time.

**Comment by codewp** [comment]
Had the same issue with dynamic properties.

My use case is to not alter product prices that are in the cart. To do it, I have added a dynamic property to the products in the cart like the below example.
```php 
function cart_loaded_from_session( $cart ) {   
    $cart_contents = $cart->get_cart();
    foreach ( $cart_contents as $cart_item_key => $cart_item ) {
	    // It is a cart item product so do not override its price.
	    $cart->cart_contents[ $cart_item_key ]['data']->is_cart_item = true;
    }
}
add_action( 'woocommerce_cart_loaded_from_session', 'cart_loaded_from_session' );
```

And somewhere else in the code, I use this dynamic property like the below example.
```php
function replace_price( $price, $product ) {
   if ( isset( $product->is_cart_item ) ) {
      return $price;
   }
   // Replace price.
}
add_action( 'woocommerce_product_get_price', 'replace_price', 10, 2 );
```

This is the best solution to detect that it is a cart item than removing filters.

Unfortunately, this code causes deprecated dynamic properties in PHP 8.2, so please let me know if you have a good solution.

Thanks

**Comment by daigo75** [comment]
@codewp I ended up using a simple strategy pattern, switching between dynamic properties in PHP 7.x and a `WeakMap` in PHP 8.x. The logic is chosen the first time an assignment or a fetching is made, so that the PHP version check is performed only once and used to instantiate a custom property tracking class, which does one of the following.

Function `set_custom_prop($object, $property, $value)` and then do the following:
* **PHP 7.x**: `$object->$property = 1234;`
* **PHP 8.x**: `$weakmap[$object]->$property = 1234;`

Function `get_custom_prop($object, $property)` and then do the following:
* **PHP 7.x**: `return $object->$property`
* **PHP 8.x**: `return $weakmap[$object]->$property;`

Of course, this is very simplified, it's just to illustrate the concept.

**Comment by codewp** [comment]
@daigo75 Thank you for your reply and for your explanation.

As I see in your example you have added dynamic property for the object inside the $weakmap and I don't know does it raise a deprecated error message or not because I didn't test it.

Is it possible for you to comment a working example of your implementation?

I have created the below example for it.
```php
class Custom_Props {

    protected $map;

    public function __construct( $map = null ) { 
        if ( null !== $map ) {
            $this->map = $map;
        } else {
            $this->map = version_compare( PHP_VERSION, '8.0', '>=' ) ? new WeakMap() : null;
        }
    }

    public function set_prop( &$object, $key, $value ) {
        if ( isset( $this->map ) ) {
            if ( ! isset( $this->map[ $object ] ) ) {
                $this->map[ $object ] = new stdClass();
            }
            $this->map[ $object ]->{ $key } = $value;
        } else {
            $object->{ $key } = $value;
        }
    }

    public function get_prop( $object, $key, $default = null ) {
        if ( isset( $this->map ) ) {
            return isset( $this->map[ $object ]->{ $key } ) ? $this->map[ $object ]->{ $key } : $default;
        }

        return isset( $object->{ $key } ) ? $object->{ $key } : $default;
    }
}

$a = new stdClass();
$custom_props = new Custom_Props();
$custom_props->set_prop( $a, 'is_cart_item', true );
$custom_props->set_prop( $a, 'extra', 'extra_data' );

var_dump( 
    $custom_props->get_prop( $a, 'is_cart_item' ), 
    $custom_props->get_prop( $a, 'extra' )
);
// outputs true, 'extra_data'
```

**Comment by daigo75** [comment]
The code I use is pretty much the same as the one you wrote. I think the only mistake is the following:
```
if ( ! isset( $this->map[ $object ] ) ) {
  $this->map[ $object ] = array();
}
```
You should initialise the element as an object, not as an array. That is:
```
if ( ! isset( $this->map[ $object ] ) ) {
  $this->map[ $object ] = new stdClass();
}
```
Then you will be able to call `$this->map[$object]->$key = $value`, and it won't raise any error (see https://php.watch/versions/8.2/dynamic-properties-deprecated#exempt).

**Comment by unfulvio-godaddy** [comment]
SkyVerge plugins also dynamically add some properties to (mainly) `WC_Order`, in a similar fashion as @daigo75 described and we were just about to open a similar issue in WC asking to add the `#[AllowDynamicProperties]` tag on top of `WC_Data`. It sounds like the easiest BC solution which wouldn't exclude implementing alternative solutions in future iterations as those being discussed in this conversation. 
`

**Comment by daigo75** [comment]
@unfulvio-godaddy based on the feedback I received so far, it's unlikely that they are going to add the `#[AllowDynamicProperties]` tag to the WC_Data class. Other classes might need the same "patch", and that will open the discussion about how many classes do need it, and how to determine that.

I've been using the `WeakMap` approach since last September and it works just fine. I implemented a strategy choice "on init", to initialise the correct data tracker (i.e. the one that tacks dynamic properties or the one using `WeakMap`) and there's almost no difference in terms of performance.

**Comment by unfulvio-godaddy** [comment]
@daigo75 that would be mostly unfortunate - I really believe it would cover majority of use cases by third party extensions since WC_Data objects are the most likely objects to be passed down by reference in hooks - which is when the properties used to be added

**Comment by unfulvio-godaddy** [comment]
Related: https://github.com/woocommerce/woocommerce/issues/35763

**Comment by rrennick** [comment]
@daigo75 Thanks for all the detailed discussion here. Earlier this year we addressed all of the open PHP 8.2 deprecation notice issues. I'll be closing #35763 this week. There are a couple new issues that are on the list to address. I will close this issue as effectively addressed. If you did want to discuss the overall approach to deprecation in upcoming versions of PHP, a [discussion thread](https://github.com/woocommerce/woocommerce/discussions) would probably be a better option.
